src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,282 @@
+            bufferedinput = new BufferedInputStream(
+                    new GZIPInputStream(bufferedinput),
+                    sizeof_bcf_header
+                    );
+        }
+
+        // try to read a BCF header
+        bufferedinput.mark(sizeof_bcf_header);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
+
+        if (bcfVersion != null) {
+            //this is BCF
+            return new BCFInputStreamIterator(bufferedinput);
+        } else {
+            //this is VCF
+            return new VCFReaderIterator(bufferedinput);
+        }
+    }
+
+    /**
+     * creates a VCF iterator from a URI It detects if the stream is a BCF
+     * stream or a GZipped stream.
+     * 
+     * @param path the Path
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final String path) throws IOException {
+        return open(ParsingUtils.openInputStream(path, null));
+    }
+
+    /**
+     * creates a VCF iterator from a Path
+     * 
+     * @param path the path
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final Path path) throws IOException {
+       return open(path, null);
+    }
+    
+    /**
+     * creates a VCF iterator from a Path
+     * 
+     * @param path the file path
+     * @param wrapper wrapper for {@link htsjdk.samtools.seekablestream.SeekablePathStream}. Can be null.
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final Path path, final Function<SeekableByteChannel, SeekableByteChannel> wrapper) throws IOException {
+        if( wrapper==null && Files.isRegularFile(path)) {// File implementation is faster
+            final File vcfFile = path.toFile();
+            /* TODO fix this when VCFFileReader will support BCF see 
+             * https://github.com/samtools/htsjdk/pull/837#discussion_r139490218
+             * https://github.com/samtools/htsjdk/issues/946
+             */
+            if(    vcfFile.getName().endsWith(IOUtil.VCF_FILE_EXTENSION) || 
+                   vcfFile.getName().endsWith(IOUtil.COMPRESSED_VCF_FILE_EXTENSION) )
+                {
+                return open(vcfFile); 
+                }
+        }
+        return open(new SeekablePathStream(path, wrapper));
+    }
+
+    
+    /**
+     * creates a VCF iterator from a File
+     * 
+     * @param file the file (can be bcf, vcf, vcf.gz)
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    @SuppressWarnings("static-method")
+    public VCFIterator open(final File file) throws IOException {
+        return new FileBasedVCFIterator(file);
+    }
+
+    /** 
+     * implementation of VCFIterator, wrapping a VCFFileReader.
+     * using this class instead of VCFReaderIterator below,
+     * avoid to wrap an InputStream into a set of BufferedInputStream 
+     * */
+    private static class FileBasedVCFIterator
```

Comment:
I'm not sure I understand the advantage of having this implementation class - why is this better than just having the open(File) method open a stream and delegate to open(InputStream). I THINK the input stream will get wrapped in (at least 1!) buffered stream either way.